posts/2021-07-29-Rust-1.54.0.md

```diff
@@ -0,0 +1,74 @@
+---
+layout: post
+title: "Announcing Rust 1.54.0"
+author: The Rust Release Team
+release: true
+---
+
+# Announcing Rust 1.54.0
+
+The Rust team is happy to announce a new version of Rust, 1.54.0. Rust is a programming language empowering everyone
+to build reliable and efficient software.
+
+If you have a previous version of Rust installed via rustup, getting Rust
+1.54.0 is as easy as:
+
+```console
+rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`][install]
+from the appropriate page on our website, and check out the
+[detailed release notes for 1.54.0][notes] on GitHub.
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: XXX
+
+## What's in 1.54.0 stable
+
+### Attributes can call macros!()
+
+Rust 1.54 supports invoking function-like macros inside attributes. One notable use case for this is including documentation from other files into Rust doc comments. For example, if your project's README represents a good documentation comment, you can use `include_str!` to directly incorporate the contents. Previously, various workarounds allowed similar functionality, but from 1.54 this is much more ergonomic.
```

Comment:
I'm uncertain if other people will find this important, but I think it might be good to briefly mention that this only works for built-in attributes of the `=EXPRESSION` form.  I don't believe it can be used for proc-macro attributes, or inputs for other styles of attributes.  Maybe something like this?  I'm not sure if "built-in" is jargon the average rust developer will understand, though.

```suggestion
Rust 1.54 supports invoking function-like macros inside built-in attributes. One notable use case for this is including documentation from other files into Rust doc comments. For example, if your project's README represents a good documentation comment, you can use `include_str!` to directly incorporate the contents. Previously, various workarounds allowed similar functionality, but from 1.54 this is much more ergonomic.
```

Reply:
I admit I was unable to find any concise user-intelligible description of what was stabilized, but comments like https://github.com/rust-lang/rust/pull/83366#issuecomment-823628221 and the PR description there (which may be up to date, not sure) suggest any expression or macro would work.

I think https://github.com/rust-lang/rust/pull/83366#issuecomment-831487787 indicates that:

* built-in attributes can get access to the expanded form (e.g., `#[path = foo!()]` gets the expansion of foo!() as input, rather than the expression `foo!()`)
* external attributes get access solely to the non-expanded form (i.e., `#[foo(bar = baz!())]` would give foo baz!() as the input token stream.
  * It's possible that external attributes also now gain access to *any* expression here -- not just macro calls -- e.g., `#[foo = 5]` would now be potentially usable, whereas before it had to be written inside a string. But I'm not quite sure on this. (No builtin macro currently takes non-string values for it's key/value pairs, so this does not apply there).

cc @rust-lang/lang @petrochenkov

Reply:
Yea, I think I may have dropped the ball a little here in terms of getting the documentation squared away.  The reference was updated (see [metaitem](https://doc.rust-lang.org/nightly/reference/attributes.html#meta-item-attribute-syntax)), but it isn't very explicit about the implications.  I didn't pursue getting it updated in detail because the reference doesn't cover expansion at all, so it can be difficult to add a small piece like this without the rest of expansion described.  

> It's possible that external attributes also now gain access to any expression here 

To clarify this statement, it's "built-in attributes also now gain". I'm not sure what is meant by "external" here, but proc-macros do not allow name=EXPR syntax.  And no built-in attribute supports non-strings.

But otherwise, yea, I think you have the gist of it.  I think the only attributes where this is really useful is `#[doc]` and `#[path]`.  The others don't seem too useful. If anyone's curious, the stable attributes that are affected are `crate_type`, `deprecated`, `doc`, `export_name`, `ignore`, `link`, `link_name`, `link_section`, `must_use`, `path`, `recursion_limit`, `should_panic`, `type_length_limit`, `windows_subsystem`.  Most of those don't benefit from early expansion.

Reply:
Niko's comment in that thread suggests otherwise - and so does the reference here https://github.com/rust-lang/reference/pull/1029/files#diff-a1074bf6cead5b27ef8c88162d6282ba3c3af77497646a7f22ac614255b26906R113

> In particular, if you have a derive (or whatever) with a custom attribute like #[foo = "bar"], you can't do:

```
#[derive(MyDerive)]
struct Foo {
    #[foo = include_str!(...)]
    x: u32
}
```

> unless you are able to pipe the include_str! through to your output.

Reply:
I think those are referring to derive macro helper attributes, which have allowed accepting arbitrary expressions since 1.30. Or maybe I'm misunderstanding what you mean?

Reply:
Or, maybe you're referring to helper attributes now allowing a macro call syntax in that position, which is new, though I'm not sure how useful that is.

Reply:
I'm not sure to be honest :) I think it'd be good to update the language in the blog post to be accurate, but I don't feel confident myself I understand enough quite yet to do so. Certainly it sounds like my interpretation of comments on the stabilization PR didn't quite match reality from what you say....

Reply:
The PR description is up to date, the relevant bit is

> Accepted syntax
The parser accepts arbitrary Rust expressions in this position, but any expression other than a macro invocation will ultimately lead to an error because it is not expected by the built-in expression forms (e.g., #[doc]). Note that decorators and the like may be able to observe other expression forms.

A more concise version is "In key-value attributes, arbitrary expressions are now allowed syntactically but not semantically; to be accepted semantically, they must expand to literal values".

> Niko's comment in that thread suggests otherwise

Niko was asking how this interacts with proc macros, in particular whether the attributes are eagerly- or lazily- expanded. The answer is they are lazily expanded.

> external attributes get access solely to the non-expanded form (i.e., #[foo(bar = baz!())] would give foo baz!() as the input token stream.
It's possible that external attributes also now gain access to any expression here -- not just macro calls -- e.g., #[foo = 5] would now be potentially usable, whereas before it had to be written inside a string.

I don't know what "external" means here, but otherwise you're correct, this only applies to top-level key-value attributes, not nested meta items. The parser doesn't treat the `x = y` in `foo(x = y)` as a key-value attribute but as an arbitrary stream of tokens. (This was discussed in more detail in the internals blog post.)

Reply:
A more concise version is "In key-value attributes, arbitrary expressions are now allowed syntactically but not semantically; to be accepted semantically, they must expand to literal values".

To be clear, this is more so saying that the new stabilization does not change the behavior of any built-in macros to accept non-string-literal values, right? Otherwise, there's not actually a restriction in some sense.

And this also means that it *is* true that proc macros are fully able to accept arbitrary Rust expressions. However, IIRC, we do validate that they must be valid Rust expressions (rather than random tokens).

Reply:
> To be clear, this is more so saying that the new stabilization does not change the behavior of any built-in macros to accept non-string-literal values, right? Otherwise, there's not actually a restriction in some sense.

This does not change the behavior of any built-in macros, but it's now checked at the macro definition site instead of in the parser itself. You could define a macro that accepted `#[x = 5]`, there just doesn't happen to be anything in the standard library that accepts that.

> And this also means that it is true that proc macros are fully able to accept arbitrary Rust expressions. However, IIRC, we do validate that they must be valid Rust expressions (rather than random tokens).

Correct.